src/Workspaces/Core/Portable/Workspace/Solution/Document.cs

```diff
@@ -475,17 +475,22 @@ internal Task<DocumentOptionSet> GetOptionsAsync(OptionSet solutionOptions, Canc
             //       snapshot model. once that is fixed, we can remove this workaround - https://github.com/dotnet/roslyn/issues/19284
             if (_cachedOptions == null)
             {
-                var newAsyncLazy = new AsyncLazy<DocumentOptionSet>(async c =>
```

Comment:
How does this prevent capturing a local? It's been moved to an instance method with the exact same code?

Reply:
➡️ In the fast path, `_cachedOptions` is not null.

Reply:
Oh, so it completely avoids allocating the lambda capturing `this`. I get it now.

Reply:
Shouldn't codegen be actually handling this case?

Reply:
I'm not sure. @dotnet/roslyn-compiler do you have more information and/or a link to an existing issue for this?

Reply:
@agocke for more insight here.

Reply:
Captured variable frames are created in the scope the variable is declared. This is an implementation detail of the compiler, so it is not documented and probably shouldn't be.

There is a feature request to use more advanced analysis to detect when the frame could be created late, but it's never something we've done before: https://github.com/dotnet/roslyn/issues/20777